core/ledger/kvledger/txmgmt/validation/validator.go

```diff
@@ -286,3 +291,58 @@ func (v *validator) validateKVReadHash(ns, coll string, kvReadHash *kvrwset.KVRe
 	}
 	return true, nil
 }
+
+type AppInitiatedPurgeUpdate struct {
+	CompositeKey              *privacyenabledstate.HashedCompositeKey
+	Version                   *version.Height
+	DeletePrivateKeyFromState bool
+}
+
+type pvtdataPurgeTracker struct {
+	m map[privacyenabledstate.HashedCompositeKey]*AppInitiatedPurgeUpdate
+}
+
+func newPvtdataPurgeTracker() *pvtdataPurgeTracker {
+	return &pvtdataPurgeTracker{
+		m: map[privacyenabledstate.HashedCompositeKey]*AppInitiatedPurgeUpdate{},
+	}
+}
+
+func (p *pvtdataPurgeTracker) update(rwset *rwsetutil.TxRwSet, version *version.Height) {
+	for _, nsRwSets := range rwset.NsRwSets {
+		for _, collHashedRwSet := range nsRwSets.CollHashedRwSets {
+			for _, hashedWrite := range collHashedRwSet.HashedRwSet.GetHashedWrites() {
+
+				ck := privacyenabledstate.HashedCompositeKey{
+					Namespace:      nsRwSets.NameSpace,
+					CollectionName: collHashedRwSet.CollectionName,
+					KeyHash:        string(hashedWrite.GetKeyHash()),
+				}
+
+				if hashedWrite.IsPurge {
+					p.m[ck] = &AppInitiatedPurgeUpdate{
+						CompositeKey:              &ck,
+						Version:                   version,
+						DeletePrivateKeyFromState: true,
+					}
+					continue
+				}
+
+				existingUpdate, ok := p.m[ck]
+				if !ok {
+					continue
+				}
+				// the hash of a key got overwritten by another transaction
+				existingUpdate.DeletePrivateKeyFromState = false
```

Comment:
This means it was purged by a prior transaction but re-written in a later transaction? Could you clarify the comment?

Reply:
Yes, that's correct. I'll expand the comment if you did not find it clear enough.